multicluster/controllers/multicluster/member/serviceexport_controller.go

```diff
@@ -491,29 +503,67 @@ func (r *ServiceExportReconciler) updateSvcExportStatus(ctx context.Context, req
-			WithEventFilter(versionChange).
+			For(&k8smcsv1alpha1.ServiceExport{}, versionChangePredicates).
+			Watches(&source.Kind{Type: &corev1.Service{}}, handler.EnqueueRequestsFromMapFunc(objectMapFunc), versionChangePredicates).
+			Watches(&source.Kind{Type: &discovery.EndpointSlice{}}, handler.EnqueueRequestsFromMapFunc(endpointSliceMapFunc), versionChangePredicates).
+			Watches(&source.Kind{Type: &mcv1alpha2.ClusterSet{}}, handler.EnqueueRequestsFromMapFunc(r.clusterSetMapFunc),
+				builder.WithPredicates(statusReadyPredicate)).
 			WithOptions(controller.Options{
 				MaxConcurrentReconciles: common.DefaultWorkerCount,
 			}).
 			Complete(r)
 	}
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&k8smcsv1alpha1.ServiceExport{}).
-		Watches(&source.Kind{Type: &corev1.Service{}}, handler.EnqueueRequestsFromMapFunc(objectMapFunc)).
-		Watches(&source.Kind{Type: &corev1.Endpoints{}}, handler.EnqueueRequestsFromMapFunc(objectMapFunc)).
-		WithEventFilter(versionChange).
+		For(&k8smcsv1alpha1.ServiceExport{}, versionChangePredicates).
+		Watches(&source.Kind{Type: &corev1.Service{}}, handler.EnqueueRequestsFromMapFunc(objectMapFunc), versionChangePredicates).
+		Watches(&source.Kind{Type: &corev1.Endpoints{}}, handler.EnqueueRequestsFromMapFunc(objectMapFunc), versionChangePredicates).
+		Watches(&source.Kind{Type: &mcv1alpha2.ClusterSet{}}, handler.EnqueueRequestsFromMapFunc(r.clusterSetMapFunc),
+			builder.WithPredicates(statusReadyPredicate)).
 		WithOptions(controller.Options{
 			MaxConcurrentReconciles: common.DefaultWorkerCount,
 		}).
 		Complete(r)
 }
 
+// clusterSetMapFunc handles ClusterSet events by enqueuing all ServiceExports
+// into the reconciler processing queue.
+func (r *ServiceExportReconciler) clusterSetMapFunc(a client.Object) []reconcile.Request {
+	clusterSet := &mcv1alpha2.ClusterSet{}
+	requests := []reconcile.Request{}
+	if a.GetNamespace() != r.namespace {
+		return requests
+	}
+
+	ctx := context.TODO()
+	err := r.Client.Get(ctx, types.NamespacedName{Namespace: a.GetNamespace(), Name: a.GetName()}, clusterSet)
+	if err == nil {
+		if len(clusterSet.Status.Conditions) > 0 && clusterSet.Status.Conditions[0].Status == corev1.ConditionTrue {
+			svcExports := &k8smcsv1alpha1.ServiceExportList{}
+			r.Client.List(ctx, svcExports)
+			existingSvcExports := sets.Set[string]{}
+			for _, svcExport := range svcExports.Items {
+				namespacedName := types.NamespacedName{
+					Name:      svcExport.GetName(),
+					Namespace: svcExport.GetNamespace(),
+				}
+				req := reconcile.Request{
+					NamespacedName: namespacedName,
+				}
+				existingSvcExports.Insert(namespacedName.String())
+				requests = append(requests, req)
+			}
+		}
+	} else if apierrors.IsNotFound(err) {
+		// Reset caches when a ClusterSet is deleted.
+		r.installedSvcs = cache.NewIndexer(svcInfoKeyFunc, cache.Indexers{})
+		r.installedEps = cache.NewIndexer(epInfoKeyFunc, cache.Indexers{})
```

Comment:
ditto, this handling will lead to no serviceExports being deleted eventually. 
EndpointExport is perhaps the same.

Reply:
Comment added.